src/backend/distributed/utils/citus_depended_object.c

```diff
@@ -0,0 +1,412 @@
+					case AggregateRelationId:
+					{
+						metaTableOid = rangeTableEntry->relid;
+						break;
+					}
+
+					default:
+					{
+						metaTableOid = InvalidOid;
+						break;
+					}
+				}
+
+				if (OidIsValid(metaTableOid))
+				{
+					/*
+					 * We do not use security quals because a postgres vanilla test fails
+					 * with a change of order for its result.
+					 */
+					query->jointree->quals = make_and_qual(
+						query->jointree->quals, CreateCitusDependentObjectExpr(varno,
+																			   metaTableOid));
+				}
+
+				MemoryContextSwitchTo(originalContext);
+			}
+		}
+
+		return query_tree_walker((Query *) node, HideCitusDependentObjectsFromPgMetaTable,
+								 context, 0);
+	}
+
+	return expression_tree_walker(node, HideCitusDependentObjectsFromPgMetaTable,
+								  context);
+}
+
+
+/*
+ * IsPgLocksTable returns true if RTE is pg_locks table.
+ */
+bool
+IsPgLocksTable(RangeTblEntry *rte)
+{
+	Oid pgLocksId = get_relname_relid("pg_locks", get_namespace_oid("pg_catalog", false));
+	return rte->relid == pgLocksId;
+}
+
+
+/*
+ * CreateCitusDependentObjectExpr constructs an expression of the form:
+ * NOT pg_catalog.is_citus_depended_object(oid, oid)
+ */
+static Node *
+CreateCitusDependentObjectExpr(int pgMetaTableVarno, int pgMetaTableOid)
+{
+	/* build the call to read_intermediate_result */
+	FuncExpr *funcExpr = makeNode(FuncExpr);
+	funcExpr->funcid = CitusDependentObjectFuncId();
+	funcExpr->funcretset = false;
+	funcExpr->funcvariadic = false;
+	funcExpr->funcformat = 0;
+	funcExpr->funccollid = 0;
+	funcExpr->inputcollid = 0;
+	funcExpr->location = -1;
+	funcExpr->args = GetFuncArgs(pgMetaTableVarno, pgMetaTableOid);
+
+	BoolExpr *notExpr = makeNode(BoolExpr);
+	notExpr->boolop = NOT_EXPR;
+	notExpr->args = list_make1(funcExpr);
+	notExpr->location = -1;
+
+	return (Node *) notExpr;
+}
+
+
+/*
+ * GetFuncArgs returns func arguments for pg_catalog.is_citus_depended_object
+ */
+static List *
+GetFuncArgs(int pgMetaTableVarno, int pgMetaTableOid)
+{
+	/*
+	 * set attribute number for the oid, which we are insterest in, inside pg meta tables.
+	 * We are accessing the 1. col(their own oid or their relation's oid) to get the related
+	 * object's oid for all of the pg meta tables except pg_enum and pg_index. For pg_enum,
+	 * class, we access its 2. col(its type's oid) to see if its type depends on citus,
+	 * so it does. For pg_index, we access its 2. col (its relation's oid) to see if its relation
+	 * depends on citus, so it does.
+	 */
+	AttrNumber oidAttNum = 1;
+	if (pgMetaTableOid == EnumRelationId || pgMetaTableOid == IndexRelationId)
+	{
+		oidAttNum = 2;
+	}
+
+	/* create const for meta table oid */
+	Const *metaTableOidConst = makeConst(OIDOID, -1, InvalidOid, sizeof(Oid),
+										 ObjectIdGetDatum(pgMetaTableOid),
+										 false, true);
+
+	/*
+	 * create a var for the oid that we are interested in,
+	 * col type should be regproc for pg_aggregate table; else oid
+	 */
```

Comment:
maybe; `Oid vartype = (pgMetaTableOid == AggregateRelationId) ? REGPROCOID : OIDOID` and use it in makeVar